### ISSUE TITLE
Query results in indefinitely pending promise when encountering JSON-RPC error 

### ISSUE THREAD

**Issue by rflechtner** [polkadot-api/polkadot-api#633]
While investigating issues around integration of PAPI and chopsticks (https://github.com/AcalaNetwork/chopsticks/issues/801) we also realised that queries can fail to reject when the chopstick client answers with a JSON-RPC error. The following code resulted in a promise that hangs indefinitely:

```js
import { createClient } from "polkadot-api"
import { getWsProvider } from "polkadot-api/ws-provider/web"
import { withPolkadotSdkCompat } from "polkadot-api/polkadot-sdk-compat"
 
const client = createClient(
    withPolkadotSdkCompat(getWsProvider('ws://localhost:8000/'))
)
const dotApi = client.getTypedApi(dot);
const accountInfo = await dotApi.query.System.Account.getValue(
    '16JGzEsi8gcySKjpmxHVrkLTHdFHodRepEz8n244gNZpr9J'
);
```

Logging of RPC messages over the wire showed the following:

```
{"jsonrpc":"2.0","id":"__INTERNAL_ID","method":"rpc_methods","params":[]}

-{"id":null,"jsonrpc":"2.0","error":{"code":-32600,"message":"Invalid JSON Request"}}
```

**Comment by josepot** [comment]
Thank you for bringing this to our attention. You've identified an issue with Chopsticks, which currently does not fully comply with the JSON-RPC specification as it only accepts numbers in its `id` field.

That said, we're committed to contributing to Chopsticks to make it library agnostic (as of today its API is tightly coupled to some PJS implementation details). We’ll prioritize this task as much as possible because we believe that Chopsticks is one of the more valuable tools in our ecosystem. While I can't give a precise timeline at this moment, our goal is to address this issue before the end of Q3 2024.

Also, [check this out](https://github.com/AcalaNetwork/chopsticks/issues/801#issuecomment-2292504132) 🙏

**Comment by rflechtner** [comment]
Hi @josepot, thanks again for working on bringing chopsticks forward by making it tool-agnostic, but there is a misunderstanding regarding the issue here. While the underlying issue that brought this to light is of course that chopsticks rejects requests that it should be handling, the fact that papi does not reject the query promise upon requests that have (conclusively) failed seems problematic to me and suggests issues with how papi manages JSON-RPC requests or, potentially, how the underlying rxjs event streams are 'promisified' (these are wild guesses, of course). 

To summarise, if papi encounters issues which mean that a state query cannot be completed, I expect to be notified via a promise rejection, instead of it silently failing with the result that the promise remains permanently unresolved - which then requires me to add a secondary timeout to these requests, something that I really shouldn't have to handle on this level.

**Comment by josepot** [comment]
The issue with PAPI not rejecting the query promise upon failed requests appears to stem from how the server (Chopsticks) is handling JSON-RPC responses. Specifically, Chopsticks is sending a response with a different `id` than the one associated with the original request, which is a violation of the JSON-RPC specification.

To clarify, if the client had received a correctly formatted error response, such as:

```json
{"id":"__INTERNAL_ID","jsonrpc":"2.0","error":{"code":-32600,"message":"Invalid JSON Request"}}
```

the promise would correctly reject as expected.

However, the client is receiving various types of messages from the server, including both notifications and responses. According to the JSON-RPC spec, the client needs to correlate the `id` of the request with the `id` of the response. Since the server is responding with an unrelated `id` (e.g., `id: null`), PAPI can't recognize that the query has failed.

In short, this isn't an issue with PAPI itself, as PAPI is sending a valid request. The problem lies with the server's non-compliant response.

**Comment by rflechtner** [comment]
Ah that makes sense. That's indeed quite the fringe edge case, which we hopefully don't see too often.
